File: cemantix.php
<?php

error_reporting(0);

class Cemantix
{
    public static $cemantix   = 'https://cemantix.certitudes.org' ;
    public static $cache_path = '~/.cemantix/';
    public static $cache      = [] ;
    public static $s_cache    = [] ;
    public static $limit      = 20 ;
    public static $padding    = 20 ;
    public static $solvers    = null ;
    public static $startDate  = "" ;
    public static $lastResp   = "" ;
    public static $num        = 0 ;
    public static $commands   = ['/help','/quit','/exit','/restart','/nearby','/history','load'];

    /**
     * @return string
     */
    private static function returnStartDate(): string
    {
        return self::$startDate ;
    }

    /**
     * @param \CurlHandle $curl
     * @return void
     */
    private static function cfgCurl($curl): void
    {
        curl_setopt_array($curl, array(
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING       => "",
            CURLOPT_MAXREDIRS      => 10,
            CURLOPT_TIMEOUT        => 3,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION   => CURL_HTTP_VERSION_1_1,
        ));

    }

    /**
     * @param string $item
     * @return mixed
     */
    private static function get($item): mixed
    {
        $curl = curl_init();
        self::cfgCurl($curl);
        curl_setopt_array($curl, array(
            CURLOPT_URL            => self::$cemantix."/".$item."?n=".self::$num,
            CURLOPT_CUSTOMREQUEST  => "GET",
        ));
        $response = curl_exec($curl);
        $ret = new stdClass();
        if (curl_errno($curl) > 0) {
            error_log("curl_error:".curl_error($curl));
            $ret->error = curl_error($curl);
        } else {
            $ret = json_decode($response);
        }
        curl_close($curl);
        return $ret ;
    }

    /**
     * @param string $action
     * @param string|null $word
     * @return mixed
     */
    private static function postWord($action, $word = null): mixed
    {
        $curl = curl_init();
        self::cfgCurl($curl);
        curl_setopt_array($curl, array(
            CURLOPT_URL            => self::$cemantix."/".$action."?n=".self::$num	,
            CURLOPT_CUSTOMREQUEST  => "POST",
            CURLOPT_HTTPHEADER     => [
                'Origin: '.self::$cemantix
            ],
            CURLOPT_POSTFIELDS     => "word=$word"
        ));
        $response = curl_exec($curl);
        $ret = new stdClass();
        if (curl_errno($curl) > 0) {
            error_log("curl_error:".curl_error($curl));
            $ret->error = "($word) ".curl_error($curl);
        } else {
            $ret = json_decode($response);
        }
        self::$lastResp = $ret ;

        // variables have been renamed
        $ret->score      = $ret->s ;
        $ret->percentile = $ret->p ;
        $ret->solvers    = $ret->v ;
        $ret->error      = $ret->e ;

        if (isset($ret->solvers)) {
            self::$solvers = $ret->solvers;
        }
        curl_close($curl);
        return $ret;
    }

    /**
     * @return void
     */
    private static function cls(): void
    {
        echo chr(27).chr(91).'H'.
             chr(27).chr(91).'J';
    }

    /**
     * @param int $num
     * @return void
     */
    private static function loadCache($num): void
    {
        self::$cache_path = str_replace(
            '~',
            getenv('HOME'),
            self::$cache_path
        );
        if (!is_dir(self::$cache_path)) {
            mkdir(self::$cache_path, 0755, true);
        }
        $filename = self::$cache_path."cem".$num.".csv" ;
        if (($handle = fopen($filename, "r")) !== false) {
            self::$cache   = [];
            while (($data = fgetcsv($handle, 1000, ",")) !== false) {
                self::$cache[$data[0]]['word']       = $data[0];
                self::$cache[$data[0]]['score']      = $data[1];
                self::$cache[$data[0]]['percentile'] = $data[2] ?? null;
                self::$cache[$data[0]]['idx']        = count(self::$cache);
            }
            self::$s_cache = self::$cache;
            usort(self::$s_cache, ['Cemantix', 'sorter']);
            fclose($handle);
        } else {
            error_log("File $filename not found") ;
        }
    }

    /**
     * @param array $a
     * @param array $b
     * @return bool
     */
    private static function sorter($a, $b): bool
    {
        if ($a['percentile'] != $b['percentile']) {
            return $b['percentile'] > $a['percentile'];
        }
        return $b['score'] > $a['score'];
    }

    /**
     * @param array $row
     * @return void
     */
    private static function writeCacheLine($row): void
    {
        $handle = fopen(self::$cache_path."cem".self::$num.".csv", "a");
        fputcsv($handle, $row);
        fclose($handle);
    }

    /**
     * @param string $str
     * @return array
     */
    private static function completeCmd($str): array
    {
        return self::$commands ;
    }

    /**
     * @param int|string $num
     * @return void
     */
    private static function loadFile($num): void
    {
        if ($num == 'today') {
            $num = self::getNum() ;
        }
        self::$num = $num ;
        self::loadCache($num);
        self::print();
    }

    /**
     * @return void
     */
    private static function init(): void
    {
        self::$startDate = date('Ymd');
        self::$num = self::getNum();
        self::loadCache(self::$num);
        self::print();
    }

    /**
     * @return int
     */
    private static function getNum(): int
    {
        $origin = new DateTime("2022-03-03");

        // Obtenir la date du jour
        $today = new DateTime();

        // Calcul du numéro du jour
        $interval = $origin->diff($today);
        $num = $interval->days + 1; // +1 car le 03/03/2022 est le jour 1

        return $num ;
    }
    /**
     *
     * str_pad handling multibyte encoding
     *
     * https://stackoverflow.com/a/14773638
     *
     * @param string $input
     * @param int $pad_length
     * @param string $pad_string
     * @param int $pad_type
     * @param string $encoding
     * @return string
     */
    private static function mb_str_pad($input, $pad_length, $pad_string = ' ', $pad_type = STR_PAD_RIGHT, $encoding = 'UTF-8'): string
    {
        $input_length = mb_strlen($input, $encoding);
        $pad_string_length = mb_strlen($pad_string, $encoding);

        if ($pad_length <= 0 || ($pad_length - $input_length) <= 0) {
            return $input;
        }

        $num_pad_chars = $pad_length - $input_length;

        switch ($pad_type) {
            case STR_PAD_RIGHT:
                $left_pad  = 0;
                $right_pad = $num_pad_chars;
                break;

            case STR_PAD_LEFT:
                $left_pad  = $num_pad_chars;
                $right_pad = 0;
                break;

            case STR_PAD_BOTH:
                $left_pad  = floor($num_pad_chars / 2);
                $right_pad = $num_pad_chars - $left_pad;
                break;
        }

        $result = '';
        for ($i = 0; $i < $left_pad; ++$i) {
            $result .= mb_substr(
                $pad_string,
                $i % $pad_string_length,
                1,
                $encoding
            );
        }
        $result .= $input;
        for ($i = 0; $i < $right_pad; ++$i) {
            $result .= mb_substr(
                $pad_string,
                $i % $pad_string_length,
                1,
                $encoding
            );
        }

        return $result;
    }

    /**
     * @param array $row
     * @param int|null $s_idx
     * @param bool $bold
     * @param int|null $solvers
     * @return void
     */
    private static function print_row($row, $s_idx = null, $bold = false, $solvers = null): void
    {
        $style = '0';
        $color = '0';
        $temperature = $row['percentile']  ;
        if ($temperature == 1000) {
            $icon = "🥳";
        }
        if ($temperature <  1000) {
            $icon = "😱";
        }
        if ($temperature <  999) {
            $icon = "🔥";
        }
        if ($temperature <  990) {
            $icon = "🥵";
        }
        if ($temperature <  900) {
            $icon = "😎";
        }
        if ($temperature <  1) {
            $icon = "🥶";
        }
        if ($temperature <  -100) {
            $icon = "🧊";
        }
        if ($bold) {
            $style = '1';
        }
        if ($row['percentile'] > 990) {
            $color = '31';
        } elseif ($row['percentile'] > 900) {
            $color = '33';
        } elseif (!empty($row['percentile'])) {
            $color = '93';
        }
        echo "\e[$style;{$color}m";
        echo sprintf(
            "\e[$style;{$color}m*\e[0m\e[{$style}m %4s",
            $row['idx']
        ) .self::mb_str_pad(
            $row['word'],
            self::$padding,
            ' ',
            STR_PAD_LEFT
        )
            .
            sprintf(
                " : %6.2f°C $icon \e[$style;{$color}m%4s\e[0m\e[{$style}m",
                $row['score'] * 1E2,
                $row['percentile'] > 0 ? $row['percentile'] : ''
            );
        if (!is_null($s_idx)) {
            printf(
                " \e[$style;{$color}m%20s\e[0m\e[{$style}m%4s/%-3s ",
                self::mb_str_pad(
                    str_repeat('◼', $row['percentile'] * 2E-2),
                    self::$padding,
                    ' ',
                    STR_PAD_RIGHT
                ),
                $s_idx + 1,
                count(self::$s_cache)
            );
        } elseif (!is_null($solvers)) {
            printf(" solvers : %s", (int) $solvers);
        }
        echo "\e[0m\n";
    }

    /**
     * @param string|null $word
     * @return void
     */
    private static function print($word = null): void
    {
        self::getScreenSize();
        self::cls();
        if ($word != null) {
            if (isset(self::$cache[$word])) {
                self::print_row(
                    self::$cache[$word],
                    null,
                    false,
                    self::$solvers
                );
                echo "\n";
            } else {
                echo strip_tags("$word\n\n");
            }
        }
        usort(self::$s_cache, ['Cemantix', 'sorter']);
        foreach (self::$s_cache as $i => $t) {
            if ($i < self::$limit) {
                self::print_row($t, $i, $word == $t['word']);
            } elseif ($word == $t['word']) {
                self::print_row($t, $i, true);
            }
        }
    }

    /**
     * @return void
     */
    private static function history(): void
    {
        $ret = self::get('history');
        self::cls();
        echo "History:\n";
        for ($i = 0; $i < self::$limit + 2; $i++) {
            $N = $ret[$i][0];
            $S = $ret[$i][1];
            $W = $ret[$i][2];
            $line = 0 + shell_exec(
                "grep 1,1000 ".self::$cache_path.
                "cem".$N.".csv -n 2>/dev/null".
                " | cut -d: -f1"
            );
            $found = ($line > 0) ? "✅" : "❌" ;
            $color = ($line > 0) ? "97" : "90" ;
            $line  = ($line == 0) ? "  " : $line ;
            echo sprintf(
                "\e[0;{$color}m*\t%4u\t%7u\t".
                self::mb_str_pad($W, 20, ' ', STR_PAD_LEFT).
                "\t%5s ".$found."\t\e[0m\n",
                $N,
                $S,
                $line
            ) ;
        }
    }

    /**
     * @return void
     */
    private static function nearby(): void
    {
        if (self::$s_cache[0]['percentile'] == 1000) {
            $ret = self::postWord('nearby', self::$s_cache[0]['word']);
            $ret = json_decode(json_encode($ret), true);
            uasort($ret, function ($a, $b) {
                return $b[0] <=> $a[0]; // Tri décroissant sur le premier élément
            });
            self::cls();
            echo "Nearby:\n";
            $i = 0 ;
            foreach ($ret as $item => $values) {
                $i++ ;

                if ($i < (self::$limit + 3)) {
                    $t = [];
                    $t['idx']        = $i;
                    $t['word']       = $item;
                    $t['score']      = $values[1] / 100;
                    $t['percentile'] = $values[0];
                    self::print_row($t);
                }
            }
        } else {
            self::print("Cheater :)");
        }
    }

    /**
     * @return void
     */
    private static function help(): void
    {
        self::cls();
        echo "
/help		You are here

/history	Prints a list of previous words.
		Depending on your terminal's size, the
		number of displayed results may vary.
/load Num	Load the game corresponding to the given
		number. To get back to current game, you can
		give its number or use «today» as a value
/nearby		Prints a list of the highest ranked words
		for the current day, in descending order of
		relevance.  You obviously need to find the word
		for this to work.  Depending on your terminal's
		size, the number of displayed results may vary.

/restart	Resets the current game, so you can start
		from scratch. (A backup is created, though)

/reset		Alias for /restart

/quit		Self explanatory

(Press Enter to return to the game)
		";
    }

    /**
     * @return void
     */
    private static function debug(): void
    {
        echo self::$lastResp;
    }

    /**
     * @return void
     */
    private static function stop(): void
    {
        exit ;
    }

    /**
     * @return void
     */
    private static function clean(): void
    {
        $num = self::get('stats')->num ;
        self::$num = $num ;
        $file = self::$cache_path."cem".$num.".csv" ;
        rename(
            $file,
            str_replace('csv', 'bak.csv', $file)
        )	;
        self::$cache   = [];
        self::$s_cache = [];
        self::start();
    }

    /**
     * @param string $cmd
     * @param string|null $params
     * @return void
     */
    private static function cmd($cmd, $params = null): void
    {
        switch ($cmd) {
            case 'nearby':
                self::nearby();
                break;
            case 'history':
                self::history();
                break;
            case 'reset':
            case 'restart':
                self::clean();
                break;
            case 'load':
                self::loadFile($params);
                break;
            case 'exit':
            case 'quit':
                self::stop();
                break;
            case 'help':
            case 'h':
            case '?':
                self::help();
                break;
            case 'debug':
                self::debug();
                break;
            case 'exec':
                eval($params.";");
                break;
            default:
                error_log("Unknown cmd <$cmd>");
        }
    }

    /**
     * @return void
     */
    private static function getScreenSize(): void
    {
        preg_match_all(
            "/rows.([0-9]+);.columns.([0-9]+);/",
            strtolower(exec('stty -a |grep columns')),
            $output
        );
        if (sizeof($output) == 3) {
            self::$limit = $output[1][0] - 5;
        }
    }

    /**
     * @return void
     */
    public static function start(): void
    {
        self::getScreenSize();
        self::init();
        readline_completion_function(['self','completeCmd']);
        while (1) {
            echo "\n";
            $word = trim(readline('> '), ' ');
            $word = preg_replace("/[^a-zA-Z\-\/\p{L}\s]`u/", "", $word);
            readline_add_history($word);
            if (preg_match('#^\/([a-z]+).*#', $word, $cmd)) {
                $params = str_replace('/'.$cmd[1].' ', '', $cmd[0]);
                self::cmd($cmd[1], $params);
            } elseif (isset(self::$cache[$word])) {
                self::print($word);
            } else {
                $ret = self::postWord('score', $word);
                // Check current date
                if (date('Ymd') > self::returnStartDate()) {
                    self::clean();
                }
                if (isset($ret->score)) {
                    self::$cache[$word]['word']       = $word;
                    self::$cache[$word]['score']      = $ret->score;
                    self::$cache[$word]['percentile'] = $ret->percentile ?? 0;
                    self::$cache[$word]['idx']        = count(self::$cache);
                    self::$s_cache[]                  = self::$cache[$word];
                    self::print($word);
                    self::writeCacheLine(
                        [$word,$ret->score,($ret->percentile ?? 0)]
                    );
                } else {
                    self::print($ret->error);
                }
            }
        }
    }
}

Cemantix::start();

File: cemantix.py
#!/usr/bin/env python

import cmd
import csv
from datetime import datetime, date
import os
import re
import signal
import sys
import configparser
import shutil
from pathlib import Path

import requests
from termcolor import colored

# Set language from command line argument
try:
    lang = sys.argv[1]

except IndexError:
    lang = "fr"
#
# Load settings from ini file config.ini where sections are selected by var lang :
config = configparser.ConfigParser()
config.read("config.ini")
if lang in config:
    settings = {key: value for key, value in config[lang].items()}
    # globals().update(settings)
    game_name = settings["game_name"]
    cemantix_url = settings["cemantix_url"]
    prefix = settings["prefix"]
    # convert string to actual date object
    origin = date.fromisoformat(settings["origin"])
else:
    raise ValueError(f"La langue {lang} n'a pas été trouvée dans le fichier config.ini")

headers = {"Origin": cemantix_url, "Referer": cemantix_url}
cachePath = os.path.expanduser("~/.cemantix/")


def handle_sigchld(signum, frame):
    while True:
        try:
            pid, _ = os.waitpid(-1, os.WNOHANG)
            if pid == 0:
                break
        except ChildProcessError:
            break


signal.signal(signal.SIGCHLD, handle_sigchld)


class Cemantix(cmd.Cmd):
    prompt = f"{game_name}> "
    intro = f"Welcome to {game_name}"
    limit = 20
    lastRow = {}
    cache = {}
    cache_idx = []
    s_cache = []
    elapsedTime = 0

    def preloop(self):
        self.cls()
        self.init()

    def precmd(self, line):
        try:
            if line[0] == "/":
                line = line[1:]
            else:
                line = "try " + line
        except IndexError:
            line = "help"
        return line

    def cls(self):
        out = os.system("cls" if os.name == "nt" else "clear")
        return out

    def init(self):
        # get today's date in Ymd format
        self.startDate = date.today()
        # no longer served
        # self.num = self.get('stats')['n']
        self.num = (self.startDate - origin).days + 1

        #  self.loadCache(self)
        self.loadCache()
        try:
            self.limit = self.getScreenSize()[0] - 3
        except ValueError:
            self.limit = 10
        self.do_printCache("")
        #  self.print()

    def print(self, word):
        """Print the word and its score"""
        self.limit = self.getScreenSize()[0] - 3
        self.cls()
        print(word)
        if word != "":
            if word in self.cache:
                self.print_row(self, self.cache[word])

        self.print_row(self, self.postWord("score", word), bold=True)

    def print_row(self, row, s_idx=0, bold=False, solvers=False):
        """Print a row of the cache"""
        # test if row is a valid dictionary
        if not isinstance(row, dict):
            row = self.lastRow

        style = ""
        color = "white"
        icon = "?"
        try:
            temperature = round(row["score"] * 1e2, 3)
        except KeyError:
            temperature = 0
        try:
            percent = row["percentile"]
        except KeyError:
            percent = 0

        icon = self.icon(percent, temperature)

        if bold:
            style = ["bold"]
        else:
            style = []
        try:
            row["percentile"] = row["percentile"] if "percentile" in row else 0
        except KeyError:
            row["percentile"] = 0
        if row["percentile"] > 990:
            color = "red"
        elif row["percentile"] > 900:
            color = "yellow"
        elif row["percentile"] > 800:
            color = "green"
        elif str(row["percentile"]) == "":
            color = "white"

        # round the score to 2 decimal after the decimal point
        try:
            row["score"] = round(row["score"] * 1e2, 4)
        except KeyError:
            row["score"] = 0

        try:
            row["word"] = row["word"]
        except KeyError:
            print(row)
            row["word"] = "Error"

        #  bargraph = "◼" * int(row['percentile'] / 50) + "◻" * (20 - int(row['percentile'] / 50))
        bargraph = "◼" * int(row["percentile"] / 50) + " " * (
            20 - int(row["percentile"] / 50)
        )
        try:
            cnt = f"{row['idx']}/{len(self.cache_idx)}"
        except KeyError:
            cnt = "0/0"
        # if solvers exists this is the current word
        if solvers:
            print(
                colored(
                    "* {:>4}{:>20} {:>6}°C{:>3}{:>5}  {:<20} {:>7} {:5.1f}ms".format(
                        " ",
                        row["word"],
                        temperature,
                        icon,
                        row["percentile"],
                        "Solvers:",
                        row["v"],
                        self.elapsedTime * 1000,
                    ),
                    color,
                    attrs=style,
                )
            )
            print("")
        else:
            try:
                idx = self.cache_idx.index(row["word"])
                thisline = "* {:>4}{:>20} {:>6}°C{:>3}{:>5} {:<20} {:>7}".format(
                    idx + 1,
                    row["word"],
                    temperature,
                    icon,
                    row["percentile"],
                    bargraph,
                    cnt,
                )
                print(colored(thisline, color, attrs=style))
            except ValueError:
                pass

    def get(self, verb):
        """Send a GET request to cemantix_url and return the response"""
        r = requests.Session()
        response = r.get(f"{cemantix_url}/{verb}?n={self.num}", headers=headers)
        return response.json()

    def post(self, verb, data):
        """Send a POST request to cemantix_url and return the response"""
        r = requests.Session()
        response = r.post(
            f"{cemantix_url}/{verb}?n={self.num}", headers=headers, data=data
        )
        self.elapsedTime = response.elapsed.total_seconds()
        return response.json()

    def getScreenSize(self):
        """return the stty rows and columns"""
        # return the stty rows and columns
        rows, columns = os.popen("stty size", "r").read().split()
        return int(rows), int(columns)

    def postWord(self, action, word):
        """
        Returns the score of the word, or an error message
        if the word is not found, or a status_code if the request
        fails
        """
        action = "score"
        data = {"word": word}
        r = requests.Session()
        out = r.post(
            f"{cemantix_url}/{action}?n={self.num}", headers=headers, data=data
        )
        self.elapsedTime = out.elapsed.total_seconds()
        if out.status_code == 200:
            return out.json()
        else:
            return out.status_code

    def icon(self, p, t):
        """
        output icon for temperature based on a value from -100 to 100
        linearized on a scale from 0 to 1000
        t : -100 to 100
        p : percentile ranging from 0 to 1000 (0: out of the top 1000 words)
            P        S
            1000 🥳 100.00
             999 😱 63.66
             990 🔥 39.41
             900 🥵 27.71
               1 😎 17.33
               0 🥶 0.00
               0  🧊 -100.00
        S scale is changing from game to game, so we can't set reliable steps
        """
        value = p
        icon = "🥶"

        if t < 0:
            p = -1
            icon = "🧊"
        if value == 1000:
            icon = "🥳"
        elif value > 998:
            icon = "😱"
        elif value > 990:
            icon = "🔥"
        elif value > 900:
            icon = "🥵"
        elif value > 1:
            icon = "😎"
        return icon

    def loadCache(self):
        num = self.num
        if not Path(cachePath).is_dir():
            os.makedirs(cachePath, mode=0o755, exist_ok=True)
        self.filename = f"{cachePath}{prefix}{num}.csv"
        #  print(self.filename)
        try:
            dataset = list()
            self.cache_idx = list()
            with open(self.filename, "r+") as file:
                csv_reader = csv.reader(file)
                for row in csv_reader:
                    if not row:
                        continue
                    dataset.append(
                        {
                            "word": row[0],
                            "score": float(row[1]),
                            "percentile": int(row[2]),
                        }
                    )
                    self.cache_idx.append(row[0])
            self.cache = dataset
            self.s_cache = sorted(
                dataset, key=lambda x: (x["score"], x["percentile"]), reverse=True
            )

        except FileNotFoundError:
            # actually no nothing
            # print(f"File {self.filename} not found", file=sys.stderr)
            pass

    def writeCacheLine(self, row):
        """Add a line to the cache file"""
        word = row[0]
        alreadyThere = any(word in d.values() for d in self.cache)
        if not alreadyThere:
            with open(self.filename, "a+", newline="") as f:
                writer = csv.writer(f)
                writer.writerow(row)

    def completedefault(self, *args):
        """Default completion function"""
        commands = set(self.completenames(*args))
        topics = set(a[5:] for a in self.get_names() if a.startswith("help_" + args[0]))
        return list(commands | topics)

    def do_init(self, line):
        self.init()

    def do_say(self, arg):
        print("You said", '"' + arg + '"')

    def do_test(self, line):
        print(self.cache)

    def do_print(self, line):
        row = self.lastRow
        row["word"] = line
        self.print_row(self, row, bold=True)

    def do_debug(self, line):
        print(f"Game #{self.num}")
        print("This is the unsorted cache:")
        print(self.cache)
        print("This is the unsorted cache index table:")
        print(self.cache_idx)
        print("This is the sorted cache:")
        print(self.s_cache)

    def do_loadFile(self, num):
        """Load the file into a dictionary"""
        self.filename = prefix + str(num) + ".csv"
        self.num = num
        # if cachePath doesn't exist, create it
        if not os.path.exists(cachePath):
            os.makedirs(cachePath)
        try:
            with open(cachePath + self.filename, "r") as f:
                reader = csv.reader(f)
                for row in reader:
                    self.cache.append(row)
        except FileNotFoundError:
            print("No cache file found")
            pass

    def do_printScreenSize(self, line):
        """Print the screen size"""
        print(self.limit)

    # define a function that prints N lines of the cache depending
    # on the number of lines the display permits
    def do_printCache(self, word):
        """Print the cache"""
        self.loadCache()
        i = 1

        for row in self.s_cache:
            try:
                row["percentile"] = row["percentile"] if "percentile" in row else 0
                row["idx"] = i
                if row["word"] == word:
                    self.print_row(row, i, solvers=False, bold=True)
                else:
                    self.print_row(row, i)
                # s_cache is the sorted cache, cache is the unsorted cache
                # so look for the current line in the unsorted cache and return
                # its index
                i += 1
            except KeyError:
                pass
            if i >= self.limit:
                break

    def do_greet(self, line):
        print("hello")

    def do_cls(self, line):
        return self.cls()

    def do_quit(self, line):
        return True

    def do_nearby(self, line):
        """
        Get nearby words
        """
        if self.s_cache[0]["percentile"] == 1000:
            word = self.s_cache[0]["word"]
            ret = self.post("nearby", {"word": word})
            sorted_data = dict(
                sorted(ret.items(), key=lambda item: item[1][0], reverse=True)
            )
            i, t = 0, {}
            self.cls()
            for key, values in sorted_data.items():
                t["idx"] = i
                t["word"] = key
                t["percentile"] = values[0]
                t["score"] = round(float(values[1]), 4)
                i += 1
                if i < (self.limit + 3):
                    # self.print_row(t, i)
                    print(
                        "* {:3} {:20} {:3} {:6.2f}°C {:4} ".format(
                            i,
                            t["word"],
                            self.icon(t["percentile"]),
                            t["score"],
                            t["percentile"],
                        )
                    )
        else:
            print("Cheater !")

    def do_reset(self, line):
        """Resets the current game"""
        # This command resets the game so the cache is flushed
        self.cache = []
        # Make a backup just in case
        shutil.move(self.filename, self.filename.replace("csv", "bak.csv"))
        self.init()

    def do_history(self, line):
        """Print the history of words and their scores"""
        self.limit = self.getScreenSize()[0] - 3
        ret = self.get("history")
        self.cls()
        print("History:")
        for i in range(0, self.limit + 2):
            Num = ret[i][0]
            Solvers = ret[i][1]
            Word = ret[i][2]
            # run grep on the cache file to get entries from 1 to 1000
            # search in given csv file for 1000
            filename = cachePath + prefix + str(Num) + ".csv"
            try:
                with open(filename, "r") as f:
                    reader = csv.reader(f)
                    foundMark = "❌"
                    idx = 0
                    NTries = ""
                    for row in reader:
                        idx = idx + 1
                        if row[0] == Word:
                            foundMark = "✅"
                            NTries = idx
                    print(
                        "* {:4} {:4} {:20} {:8} {:1}".format(
                            Num, NTries, Word, Solvers, foundMark
                        )
                    )
            except FileNotFoundError:
                pass

    def do_try(self, word):
        """
        Send word via POST request to cemantix_url and return its score
        """
        # If the date is different from the one in the cache, we have to re-initialiaze the game
        if self.startDate != date.today():
            self.init()
        self.limit = self.getScreenSize()[0] - 3
        #   self.cls()
        self.loadCache()
        response = self.postWord("score", word)

        self.cls()
        try:
            response["percentile"] = response["p"]
        except KeyError:
            response["percentile"] = 0

        try:
            response["score"] = round(response["s"], 4)
        except KeyError:
            response["score"] = 0

        rcode = "Ok"

        try:
            response["error"] = response["e"]
            rcode = "Error"
            print(re.sub("<[^<]+?>", "", response["error"]))
            print()
        except KeyError:
            pass

        self.lastRow = response
        self.lastRow["word"] = word
        if rcode == "Ok":
            try:
                if int(response["percentile"]) == 1000:
                    print("Gagné !")

                rcode = "Ok"
            except KeyError:
                response["percentile"] = 0
                rcode = "Error"
            row = [word, response["score"], response["percentile"]]
            self.writeCacheLine(row)

        #  s_idx = 1
        self.print_row(self, self.lastRow, 0, response["score"])
        self.do_printCache(word)


if __name__ == "__main__":
    Cemantix().cmdloop()
